don't override document.title if set to None

chriddyp opened the pull request:
As requested by a [Dash Enterprise](https://plotly.com/dash) customer.

This allows users to override `document.title` during runtime with a
clientside callback or a special component
(https://github.com/plotly/dash-core-components/issues/833)

Example of modifying during runtime:
```python
import dash
import dash_html_components as html
import dash_core_components as dcc
from dash.dependencies import Input, Output

app = dash.Dash(__name__, update_title=None)


app.layout = app.layout = html.Div([
    dcc.Input(id='input'),
    html.Div(id='output'),

    html.Div(id='dummy'),
    dcc.Tabs(id='tabs-example', value='tab-1', children=[
        dcc.Tab(label='Tab one', value='tab-1'),
        dcc.Tab(label='Tab two', value='tab-2'),
    ]),
    html.Div(id='tabs-example-content')
])


app.clientside_callback(
    """
    function(tab_value) {
        console.log(tab_value);
        document.title = tab_value;
        return null; // dummy output
    }
    """,
    Output('dummy', 'children'),
    [Input('tabs-example', 'value')]
)


@app.callback(Output('output', 'children'), [Input('input', 'value')])
def update_output(value):
    return value

if __name__ == '__main__':
    app.run_server(debug=True)
```

Example with custom table title via index string:
```python
import dash
import dash_html_components as html
import dash_core_components as dcc
from dash.dependencies import Input, Output

app = dash.Dash(__name__, update_title=None)

app.index_string = '''
<!DOCTYPE html>
<html>
    <head>
        {%metas%}
        <title>Hello World</title>
        {%favicon%}
        {%css%}
    </head>
    <body>
        <div>My Custom header</div>
        {%app_entry%}
        <footer>
            {%config%}
            {%scripts%}
            {%renderer%}
        </footer>
        <div>My Custom footer</div>
    </body>
</html>
'''


app.layout = app.layout = html.Div([
    dcc.Input(id='input'),
    html.Div(id='output'),

    html.Div(id='dummy'),
    dcc.Tabs(id='tabs-example', value='tab-1', children=[
        dcc.Tab(label='Tab one', value='tab-1'),
        dcc.Tab(label='Tab two', value='tab-2'),
    ]),
    html.Div(id='tabs-example-content')
])


@app.callback(Output('output', 'children'), [Input('input', 'value')])
def update_output(value):
    return value

if __name__ == '__main__':
    app.run_server(debug=True)
```

 This is a follow up of https://github.com/plotly/dash/pull/1315/.

Marc-Andre-Rivet commented on dash-renderer/src/components/core/DocumentTitle.react.js:
```diff
@@ -7,16 +7,13 @@ class DocumentTitle extends Component {
     UNSAFE_componentWillReceiveProps(props) {
         if (this.state.update_title && props.isLoading) {
             document.title = this.state.update_title;
-        } else {
```
While this allows the clientside callback to set the value, it seems to break existing tests / behavior. Would a better verification be to online revert back to `inititalTitle` if the value is `update_title` instead?

chriddyp replied:
Thanks! I totally misinterpreted what `update_title` meant. I believe https://github.com/plotly/dash/pull/1343/commits/e190cc941b89275677f0b3b995d21d5220e0209b is the logic we're looking for

alexcjohnson replied:
I think it'd be safer to bail out entirely (before even the `props.isLoading` test) if `!update_title`. Then this entire component would be a noop, as it should be in that case; callbacks can do whatever they want with `document.title` and we won't interfere.

When you _do_ have an `update_title`, I may agree with @Marc-Andre-Rivet that there needs to be a test when we're reverting back to the non-updating title. Consider a callback chain, where the first callback changes `document.title`, but we're still in a loading state so `DocumentTitle` doesn't receive new props and we don't pick up the changed title into `this.state`. Then after the second callback we reset `document.title` to `this.state.title`, losing the callback-provided title.

Seems to me the logic you have right now is fine during loading, but when reverting to `this.state.title` below it would be better to do something like:
```js
if (document.title === this.state.update_title) {
    document.title = this.state.title;
} else {
    this.setState({title: document.title});
}
```

chriddyp replied:
Thanks @alexcjohnson @Marc-Andre-Rivet ! I believe I have addressed your comments and added the appropriate tests now.

chriddyp commented:
Huh, odd that pylint would fail. I didn't change any python files.
![image](https://user-images.githubusercontent.com/1280389/88353641-29f1b300-cd13-11ea-80d5-fcb57765b00a.png)

alexcjohnson commented:
Tangentially related to this PR - you can set `app.title` to control the regular page title (which I totally forgot about when reviewing #1315). We use this internally, but I don't see it documented anywhere, and it feels inconsistent with the `update_title` kwarg. Perhaps we should support `title` as a kwarg too? Also `app.config.update_title` works, so it would be nice if `app.config.title` did too. We could go full symmetry and allow `app.update_title`, but it might be cleaner to deprecate `app.title` and encourage both of these to go through kwargs or `app.config`.

alexcjohnson commented:
> Huh, odd that pylint would fail.

weird, perhaps we somehow got `isort` v5? Maybe we need to pin it to `4.3.21` in `requires_dev.txt`? `pylint` tries to handle that itself, but perhaps someone else brings it in unpinned first?

chriddyp commented:
> Perhaps we should support title as a kwarg too? 

I added this as well and kept the `app.title` for backwards compatibility. Also added a unit test here.

chriddyp commented:
Huh, seeing these re-introduced snapshots on Percy

![image](https://user-images.githubusercontent.com/1280389/88560104-8ccc9e00-cfe2-11ea-9b78-c513a5eb4626.png)

alexcjohnson commented on dash/dash.py:
```diff
@@ -725,7 +739,9 @@ def index(self, *args, **kwargs):  # pylint: disable=unused-argument
         renderer = self._generate_renderer()
-        title = getattr(self, "title", "Dash")
+
+        # use self.title instead of app.config.title for backwards compatibility
```
In principle we could make title into a `@property` that syncs `app.title` with `app.config.title` - but not a big deal. Most people should just use the kwarg regardless.

alexcjohnson commented:
Dunno what the deal is with the "reintroduced" percy snapshots - but I'd just ignore it this time and look into it if we see that again.

OverLordGoldDragon commented:
Isn't below simpler? 

```python
def interpolate_index(self, **kwargs):
    kwargs['title'] = 'Custom Title'
    return dash.Dash.interpolate_index(self, **kwargs)
app.interpolate_index = interpolate_index.__get__(app)
```

alexcjohnson commented:
I'm not sure what you mean @OverLordGoldDragon - simpler than `app = dash.Dash(title='Custom Title')`?

OverLordGoldDragon commented:
@alexcjohnson Surely not, but your code blob isn't shown in this PR nor in Releases (it was unclear where the "`title` parameter" should go) - thanks for clarifying.

alexcjohnson commented:
OIC - yeah, this PR expanded a bunch from its original purpose, so we never showed the simple examples - sorry for the confusion. We've documented the new behavior at https://dash.plotly.com/external-resources (though @chriddyp it feels to me a bit hidden on that page - maybe time for a dedicated `dash.Dash` class API reference page?)
As we've discussed, this is a big argument in favor of merging the docs into this repo - then https://github.com/plotly/dash-docs/pull/931 would have been part of this PR.

hivemall commented:
VERY much appreciated feture + doc well, awesome
